src/util/locking/MultiThreadedResourceLocker.ts

```diff
@@ -0,0 +1,131 @@
+import { createHash } from 'crypto';
+import { ensureDirSync, pathExists, readdir, rmdir } from 'fs-extra';
+import type { LockOptions, UnlockOptions } from 'proper-lockfile';
+import { lock, unlock } from 'proper-lockfile';
+import type { ResourceIdentifier } from '../../http/representation/ResourceIdentifier';
+import type { Finalizable } from '../../init/final/Finalizable';
+import { getLoggerFor } from '../../logging/LogUtil';
+import { InternalServerError } from '../errors/InternalServerError';
+import type { AttemptSettings } from '../LockUtils';
+import { wait } from '../LockUtils';
+import { ensureTrailingSlash, joinFilePath } from '../PathUtil';
+import type { ResourceLocker } from './ResourceLocker';
+
+const defaultLockOptions: LockOptions = {
+  realpath: false,
+  retries: 0,
+};
+
+const defaultUnlockOptions: UnlockOptions = {
+  realpath: false,
+};
+
+const attemptDefaults: Required<AttemptSettings> = { retryCount: -1, retryDelay: 50, retryJitter: 30 };
+
+/** Directory to store the locks in. */
+const LOCK_DIR = '/.internal/locks';
+
+/**
+ * A resource locker making use of the `proper-lockfile` library.
+ * Note that no locks are kept in memory.
+ */
+export class MultiThreadedResourceLocker implements ResourceLocker, Finalizable {
+  protected readonly logger = getLoggerFor(this);
+  private readonly attemptSettings: Required<AttemptSettings>;
+  /** Folder that stores the locks */
+  private readonly lockFolder: string;
+
+  public constructor(rootFilePath: string, attemptSettings: AttemptSettings = {}) {
+    this.attemptSettings = { ...attemptDefaults, ...attemptSettings };
+    this.lockFolder = ensureTrailingSlash(joinFilePath(rootFilePath, ...LOCK_DIR.split('/')));
+    ensureDirSync(this.lockFolder);
+  }
+
+  public async acquire(identifier: ResourceIdentifier): Promise<void> {
+    const { path } = identifier;
+    this.logger.debug(`Acquiring lock for ${path}`);
+    try {
+      const opt = this.generateOptions(identifier, defaultLockOptions);
+      await this.tryFn((): Promise<() => Promise<void>> => lock(path, opt));
+    } catch (err: unknown) {
+      throw new InternalServerError(`Error trying to acquire lock for ${path}. ${err}`);
+    }
+  }
+
+  public async release(identifier: ResourceIdentifier): Promise<void> {
+    const { path } = identifier;
+    this.logger.debug(`Releasing lock for ${path}`);
+
+    try {
+      const opt = this.generateOptions(identifier, defaultUnlockOptions);
+      await this.tryFn((): Promise<void> => unlock(path, opt));
```

Comment:
I'm assuming this only fails in the case of disk errors? Does the library provide any support for these cases? Since I saw a `retries` option at the top of this file.